encoding/dot/decode_test.go

```diff
@@ -0,0 +1,225 @@
+	// Node definitions.
+	0 [label="foo 2"];
+	1 [label="bar 2"];
+
+	// Edge definitions.
+	0 -- 1 [label="baz 2"];
+}`
+
+// Below follows a minimal implementation of a graph capable of validating the
+// round-trip encoding and decoding of DOT graphs with nodes and edges
+// containing DOT attributes.
+
+// DirectedGraph extends simple.DirectedGraph to add NewNode and NewEdge
+// methods for creating user-defined nodes and edges.
+//
+// DirectedGraph implements the dot.Builder interface.
+type DirectedGraph struct {
+	*simple.DirectedGraph
+}
+
+// newDirectedGraph returns a new directed capable of creating user-defined
+// nodes and edges.
+func newDirectedGraph() *DirectedGraph {
+	return &DirectedGraph{
+		DirectedGraph: simple.NewDirectedGraph(0, 0),
+	}
+}
+
+// NewNode adds a new node with a unique node ID to the graph.
+func (g *DirectedGraph) NewNode() graph.Node {
+	n := &Node{
+		Node: simple.Node(g.NewNodeID()),
+	}
+	g.AddNode(n)
+	return n
+}
+
+// NewEdge adds a new edge from the source to the destination node to the graph,
+// or returns the existing edge if already present.
+func (g *DirectedGraph) NewEdge(from, to graph.Node) graph.Edge {
+	if e := g.Edge(from, to); e != nil {
+		return e
+	}
+	e := &Edge{
+		Edge: simple.Edge{F: from, T: to},
+	}
+	g.SetEdge(e)
+	return e
+}
+
+// UndirectedGraph extends simple.UndirectedGraph to add NewNode and NewEdge
+// methods for creating user-defined nodes and edges.
+//
+// UndirectedGraph implements the dot.Builder interface.
+type UndirectedGraph struct {
+	*simple.UndirectedGraph
+}
+
+// newUndirectedGraph returns a new undirected capable of creating user-defined
+// nodes and edges.
+func newUndirectedGraph() *UndirectedGraph {
+	return &UndirectedGraph{
+		UndirectedGraph: simple.NewUndirectedGraph(0, 0),
+	}
+}
+
+// NewNode adds a new node with a unique node ID to the graph.
+func (g *UndirectedGraph) NewNode() graph.Node {
+	n := &Node{
+		Node: simple.Node(g.NewNodeID()),
+	}
+	g.AddNode(n)
+	return n
+}
+
+// NewEdge adds a new edge from the source to the destination node to the graph,
+// or returns the existing edge if already present.
+func (g *UndirectedGraph) NewEdge(from, to graph.Node) graph.Edge {
+	if e := g.Edge(from, to); e != nil {
+		return e
+	}
+	e := &Edge{
+		Edge: simple.Edge{F: from, T: to},
+	}
+	g.SetEdge(e)
+	return e
+}
+
+// Node extends simple.Node with a label field to test round-trip encoding and
+// decoding of node DOT label attributes.
+type Node struct {
+	simple.Node
+	// Node label.
+	Label string
+}
+
+// UnmarshalDOTAttr decodes a single DOT attribute.
+func (n *Node) UnmarshalDOTAttr(attr dot.Attribute) error {
+	switch attr.Key {
+	case "label":
+		n.Label = attr.Value
+	default:
+		return fmt.Errorf("unable to unmarshal node DOT attribute with key %q", attr.Key)
+	}
+	return nil
+}
+
+// DOTAttributes returns the DOT attributes of the node.
+func (n *Node) DOTAttributes() []dot.Attribute {
+	var attrs []dot.Attribute
```

Comment:
Same rationale here, if users wanted to add more than one attribute, it is easy to extend. I'll update the code of the test case. In the example I'll add another attribute and keep the current style.